File: lib/iscp/messages/SubwooferLevelCommandMsg.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

import "../EISCPMessage.dart";
import "../ISCPMessage.dart";
import "EnumParameterMsg.dart";

enum SubwooferLevelCommand
{
    NONE,
    UP,
    DOWN
}

/*
 * Subwoofer (temporary) Level Command
 */
class SubwooferLevelCommandMsg extends ISCPMessage
{
    static const String CODE = "SWL";

    static const String KEY = "Subwoofer Level";
    static const int NO_LEVEL = 0xFF;

    static const ExtEnum<SubwooferLevelCommand> ValueEnum = ExtEnum<SubwooferLevelCommand>([
        EnumItem(SubwooferLevelCommand.NONE, defValue: true),
        EnumItem(SubwooferLevelCommand.UP),
        EnumItem(SubwooferLevelCommand.DOWN)
    ]);

    late EnumItem<SubwooferLevelCommand> _command;
    int _level = NO_LEVEL;
    int _cmdLength = NO_LEVEL;

    SubwooferLevelCommandMsg(EISCPMessage raw) : super(CODE, raw)
    {
        _command = ValueEnum.defValue;
        _level = ISCPMessage.nonNullInteger(getData, 16, NO_LEVEL);
        _cmdLength = getData.length;
    }

    SubwooferLevelCommandMsg.output(SubwooferLevelCommand v) :
            super.output(CODE, ValueEnum.valueByKey(v).getCode)
    {
        _command = ValueEnum.valueByKey(v);
        _level = NO_LEVEL;
        _cmdLength = NO_LEVEL;
    }

    SubwooferLevelCommandMsg.value(int level, int cmdLength) :
            super.output(CODE, _getParameterAsString(level, cmdLength))
    {
        _command = ValueEnum.defValue;
        _level = level;
        _cmdLength = cmdLength;
    }

    static String _getParameterAsString(int level, int cmdLength)
    {
        if (level == 0)
        {
            return cmdLength == 2 ? "00" : "000";
        }
        else
        {
            final String s = level < 0 ? "-" : "+";
            final String format = cmdLength == 2 ?
                level.abs().toRadixString(16) : level.abs().toRadixString(16).padLeft(2, '0');
            return s + format.toUpperCase();
        }
    }

    int get getLevel
    => _level;

    int get getCmdLength
    => _cmdLength;

    @override
    String toString()
    => super.toString() + "[COMMAND=" + _command.toString()
        + "; LEVEL=" + _level.toString()
        + "; CMD_LENGTH=" + _cmdLength.toString() + "]";

    @override
    bool hasImpactOnMediaList()
    {
        return false;
    }
}

File: app/src/main/java/com/mkulesh/onpc/iscp/messages/SubwooferLevelCommandMsg.java
/*
 * Enhanced Controller for Onkyo and Pioneer
 * Copyright (C) 2018-2021 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.iscp.messages;

import com.mkulesh.onpc.iscp.EISCPMessage;
import com.mkulesh.onpc.iscp.ISCPMessage;
import com.mkulesh.onpc.utils.Utils;

import androidx.annotation.NonNull;

/*
 * Subwoofer (temporary) Level Command
 */
public class SubwooferLevelCommandMsg extends ISCPMessage
{
    public final static String CODE = "SWL";

    public final static String KEY = "Subwoofer Level";
    public final static int NO_LEVEL = 0xFF;
    private int level, cmdLength;

    SubwooferLevelCommandMsg(EISCPMessage raw) throws Exception
    {
        super(raw);
        try
        {
            level = Integer.parseInt(data, 16);
            cmdLength = data.length();
        }
        catch (Exception e)
        {
            level = NO_LEVEL;
            cmdLength = NO_LEVEL;
        }
    }

    public SubwooferLevelCommandMsg(int level, int cmdLength)
    {
        super(0, null);
        this.level = level;
        this.cmdLength = cmdLength;
    }

    public int getLevel()
    {
        return level;
    }

    public int getCmdLength()
    {
        return cmdLength;
    }

    @NonNull
    @Override
    public String toString()
    {
        return CODE + "[" + data + "; LEVEL=" + level + "; CMD_LENGTH=" + cmdLength + "]";
    }

    @Override
    public EISCPMessage getCmdMsg()
    {
        return new EISCPMessage(CODE, Utils.intLevelToString(level, cmdLength));
    }

    @Override
    public boolean hasImpactOnMediaList()
    {
        return false;
    }
}
